securityscanutils/securityscan.go

```diff
@@ -0,0 +1,284 @@
+package securityscanutils
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/imroc/req"
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/githubutils"
+	"github.com/solo-io/go-utils/log"
+	"github.com/solo-io/go-utils/osutils"
+)
+
+type SecurityScanner struct {
+	Repos []*SecurityScanRepo
+}
+
+type SecurityScanRepo struct {
+	Repo  string
+	Owner string
+	Opts  *SecurityScanOpts
+}
+
+type SecurityScanOpts struct {
+	// The following directory structure will be created in your output dir.
+	/*
+	   OUTPUT_DIR/
+	   ├─ markdown_results/
+	   │  ├─ repo1/
+	   │  │  ├─ 1.4.12/
+	   │  │  ├─ 1.5.0/
+	   │  ├─ repo2/
+	   │  │  ├─ 1.4.13/
+	   │  │  ├─ 1.5.1/
+	   ├─ sarif_results/
+	   │  ├─ repo1/
+	   │  │  ├─ 1.4.12/
+	   │  │  ├─ 1.5.0/
+	   │  ├─ repo2/
+	   │  │  ├─ 1.4.13/
+	   │  │  ├─ 1.5.1/
+	*/
+	OutputDir string
+	// A mapping of version constraints to images scanned.
+	// If 1.6 had images "gloo", "discovery" and 1.7 introduced a new image "rate-limit",
+	// the map would look like:
+	/*
+	   ' >= 1.6': ["gloo", "discovery"]
+	   ' >= 1.7': ["gloo", "discovery", "rate-limit"]
+	*/
+	// where the patch number is explicitly not set so that these versions can match all
+	// 1.6.x-x releases
+	ImagesPerVersion map[string][]string
+	// VersionConstraint on releases to security scan
+	// any releases that do not pass this constraint will not be security scanned.
+	// If left empty, all versions will be scanned
+	VersionConstraint *semver.Constraints
+
+	// Required: image repo (quay.io, grc.io, gchr.io)
+	ImageRepo string
+
+	// Uploads Sarif file to github security code-scanning results
+	// e.g. https://github.com/solo-io/gloo/security/code-scanning
+	UploadCodeScanToGithub bool
+}
+
+// Main method to call on SecurityScanner which generates .md and .sarif files
+// in OutputDir as defined above per repo. If UploadCodeScanToGithub is true,
+// sarif files will be uploaded to the repository's code-scanning endpoint.
+func (s *SecurityScanner) GenerateSecurityScans(ctx context.Context) error {
+
+	client, err := githubutils.GetClient(ctx)
+	if err != nil {
+		return eris.Wrap(err, "error initializing github client")
+	}
+	for _, repo := range s.Repos {
+		opts := repo.Opts
+		allReleases, err := githubutils.GetAllRepoReleases(ctx, client, repo.Owner, repo.Repo)
+		if err != nil {
+			return eris.Wrapf(err, "unable to fetch all github releases for github.com/%s/%s", repo.Owner, repo.Repo)
+		}
+		githubutils.SortReleasesBySemver(allReleases)
```

Comment:
we call sort and then filteredReleases. (1) are we assured that the ordering is maintained in the new list? (2) we could reduce the number of items that are sorted by sorting only on the filtered releases.